C compilation error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.11 21d9a5d, press to see full `v doctor` output</summary>

|V full version      |V 0.4.11 21d9a5ded57a692692cc5612ddeb181482fc0975

[custom-set.txt](https://github.com/user-attachments/files/21702681/custom-set.txt)
[run_test.txt](https://github.com/user-attachments/files/21702680/run_test.txt)

|OS                  |linux, N/A
|Processor           |16 cpus, 64bit, little endian, AMD Ryzen 7 7730U with Radeon Graphics
|Memory              |12.62GB/15.02GB
|                    |
|V executable        |/usr/local/v/v
|V last modified time|2025-08-05 10:22:09
|                    |
|V home dir          |OK, value: /usr/local/v
|VMODULES            |OK, value: /root/.vmodules
|VTMP                |OK, value: /tmp/v_0
|Current working dir |OK, value: /home/kah/workspace/exercism-vlang/exercises/practice/custom-set
|                    |
|Git version         |git version 2.39.2
|V git status        |N/A
|.git/config present |false
|                    |
|cc version          |cc (Debian 12.2.0-14) 12.2.0
|gcc version         |gcc (Debian 12.2.0-14) 12.2.0
|clang version       |N/A
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |N/A
|emcc version        |N/A
|glibc version       |ldd (Debian GLIBC 2.36-9+deb12u7) 2.36

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  run_test.v && run_test`

or

`v test .`

The source files are attached as `custom-set.txt` and `run_test.txt` (GitHub doesn't seem to like the `.v` extension, so I renamed to use `.txt`).

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_0/run_test.01K28J7VCJ8R5NQVBXVBV0TFJG.tmp.c:1853: error: invalid type
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Compilation to succeed, without error

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -848,6 +848,15 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
 					c.error('cannot assign `nil` to struct field `${init_field.name}` with type `${expected_type_sym.name}`',
 						init_field.expr.pos.extend(init_field.expr.expr.pos()))
 				}
+				if mut init_field.expr is ast.CallExpr
+					&& init_field.expr.return_type.has_flag(.generic) {
+					expected_type := c.unwrap_generic(init_field.expected_type)
+					got_type_ret := c.unwrap_generic(init_field.expr.return_type)
+					if expected_type != got_type_ret {
+						c.error('cannot assign `${c.table.type_to_str(got_type_ret)}` to struct field `${init_field.name}` with type `${c.table.type_to_str(expected_type)}`',
+							init_field.expr.pos)
+					}
+				}
 			}
 			if !node.has_update_expr {
 				c.check_uninitialized_struct_fields_and_embeds(node, type_sym, mut info, mut

--- a/vlib/v/checker/tests/generic_field_init_with_call_err.out
+++ b/vlib/v/checker/tests/generic_field_init_with_call_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generic_field_init_with_call_err.vv:12:10: error: cannot assign `map[int]int` to struct field `items` with type `map[int]u8`
+   10 | pub fn CustomSet.new[T](elements []T) CustomSet[T] {
+   11 |     return CustomSet[T]{
+   12 |         items: to_map[T, int, T, int](invert[T, int](from_array[T](elements)), fn [T](key T, _ int) (T, int) {
+      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   13 |             return key, 10000
+   14 |         })

--- a/vlib/v/checker/tests/generic_field_init_with_call_err.vv
+++ b/vlib/v/checker/tests/generic_field_init_with_call_err.vv
@@ -1,0 +1,18 @@
+module main
+
+import maps { from_array, invert, to_map }
+
+struct CustomSet[T] {
+mut:
+	items map[T]u8
+}
+
+pub fn CustomSet.new[T](elements []T) CustomSet[T] {
+	return CustomSet[T]{
+		items: to_map[T, int, T, int](invert[T, int](from_array[T](elements)), fn [T](key T, _ int) (T, int) {
+			return key, 10000
+		})
+	}
+}
+
+dump(CustomSet.new([1]))

--- a/vlib/v/checker/tests/generics_struct_init_with_inconsistent_generic_types_3_err.out
+++ b/vlib/v/checker/tests/generics_struct_init_with_inconsistent_generic_types_3_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generics_struct_init_with_inconsistent_generic_types_3_err.vv:12:25: error: cannot assign `string` to struct field `val` with type `int`
+   10 | pub fn (me Maybe[T]) map[T, U](f fn (T) U) Maybe[U] {
+   11 |     if me.present {
+   12 |         return Maybe[U]{true, f(me.val)}
+      |                               ~~~~~~~~~
+   13 |     }
+   14 |     return Maybe[U]{}

--- a/vlib/v/checker/tests/generics_struct_init_with_inconsistent_generic_types_3_err.vv
+++ b/vlib/v/checker/tests/generics_struct_init_with_inconsistent_generic_types_3_err.vv
@@ -22,7 +22,7 @@ pub fn (o Maybe[T]) get() ?T {
 	}
 }
 
-fn test_generics_struct_init_with_inconsistent_generic_types() {
+fn main() {
 	m := some[int](12)
 	ret := m.map[int, string](fn (i int) string {
 		return i.str()

